Fix incorrect index->reltuples after VACUUM

haolinw opened the pull request:
Commit 1: fix unpredictable behavior during AO index vacuum
This is a backport of 6X https://github.com/greenplum-db/gpdb/pull/14260

Commit 2: Fix incorrect index->reltuples after VACUUM
Given the scenario of updating stats during VACUUM:

1) coordinator vacuums and updates stats of its own;
2) then coordinator dispatches vacuum to segments;
3) coordinator combines stats received from segments
   to overwrite the stats of its own.

Because upstream introduced a feature which could skip
full index scan uring cleanup of B-tree indexes when
possible (refer to: https://github.com/postgres/postgres/commit/857f9c36cda520030381bd8c2af20adf0ce0e1d4),
there was a case in QD-QEs distributed deployment that
some QEs could skip full index scan and stop updating
statistics, resulting in QD being unable to collect all
QEs' stats thus overwrote a paritial accumulated value
to index->reltuples. More interesting, it usually happened
starting from the 3rd time of consecutively VACUUM after
fresh inserts due to above skipping index scan criteria.

This patch is intended to prevent from above issue by two aspects:
on QE, do not skip full index scan, report current statistics
to QD as requested;
on QD, restrict updating statistics only when collecting all QEs'
data completely, if the reporting QE number doesn't match
the total number of dispatched QEs, no stats update happens.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

yaowangm reviewed (commented):
There are still something I have not understood completely. I will continue tomorrow

soumyadeep2007 commented on src/backend/access/nbtree/nbtree.c:
```diff
@@ -890,6 +890,10 @@ _bt_vacuum_needs_cleanup(IndexVacuumInfo *info)
 	bool		result = false;
 
+	/* Return true directly on QE for stats collection from QD. */
+	if (gp_vacuum_needs_update_stats())
```
Agree with Yao since there is only one use of this function.

haolinw replied:
Well I think coding such pattern like if (Gp_role == GP_ROLE_EXECUTE) all over the code is really bad. Although we have moved so far in this way but I still think limiting it as few as possible. Ideally, the determination of whether it is QD or QE could happen in few higher layer (e.g. at or above executor). Having this check across multiple layers easily breaks the expected behavior of current context, also it easily hide cases if not going deep until reaching to this check.
Actually, I did try coding like that at first but it needed to include header file containing role's definition which I feel really not necessary. Based on that, I would like to keep this short function and may define it as a static inline function in vacuum.h to reduce the overhead of calling it (understand it is not a big deal).

yaowangm replied:
Yes, it makes sense to avoid using `Gp_role == GP_ROLE_EXECUTE`. But can we use a common inline function with general-purpose name `like IS_QUERY_EXECUTOR_BACKEND()` instead of a couple of functions with same logic and different specific names?

Aegeaner replied:
Or we can introduce a macro for this.

soumyadeep2007 reviewed (commented):
Mainly reviewed the 1st commit so far. 

Suggestions:
(1) Please reword the commit message as follows:
```
vacuum_appendonly_index: Fix IndexVacuumInfo initialization

* IndexVacuumInfo.analyze_only was not initialized, leading to possible
no-ops in index_vacuum_cleanup(), if the garbage value for it was
non-zero.

* IndexVacuumInfo.num_heap_tuples should be set to the AO table's old
reltuples value (value prior to kicking off vacuum). Please see
IndexVacuumInfo and lazy_vacuum_index() for details. As explained in
the code comments, num_heap_tuples will always be an estimate during
ambulkdelete(). Since dedd4075c9f, we have been using the index rel's
reltuples instead. This is wrong (eg in scenarios where the index has
been created post data manipulation). So, correctly use the table's
reltuples value to provide the estimate.
``` 
(2) Is the 4th commit: `add test into btree_index.sql
` a test for the 2nd commit: `Fix incorrect index->reltuples after VACUUM
`?

haolinw commented on src/include/cdb/cdbdispatchresult.h:
```diff
@@ -29,6 +29,7 @@ typedef struct CdbPgResults
 {
 	struct pg_result **pg_results;
 	int numResults;
+	int numDispatches;
```
Done.

haolinw commented on src/include/commands/vacuum.h:
```diff
@@ -206,6 +206,16 @@ typedef struct VPgClassStats
 	BlockNumber relallvisible;
 } VPgClassStats;
 
+typedef struct VPgClassStatsCombo
```
Don't want to change the statistic transporting protocol, especially for ensuring upgrade running properly. Given VPgClassStats was used for messaging of statistics between QD and QE, QD could be unable to differentiate which QE was upgraded with the new structure during some upgrade scenario.

haolinw replied:
Actually the new introduced field `count` doesn't need to be added into message at all because it is just used for counting on QD side. It is meaningless on QE so don't think it's necessary to change the current messaging structure of VPgClassStats. Just treat VPgClassStatsCombo as a local helper structure for accumulating the statistics  only for QD. So don't see any improvement is necessary here, or could you list some example to explain what needs TODO ?

yaowangm replied:
I meant you can use `VPgClassStats` as a member of `VPgClassStatsCombo` to avoid duplication code, like:

```
typedef struct VPgClassStatsCombo
{
    VPgClassStats pgClassStats;
    int count;
}
```
But it depends on you.

Aegeaner replied:
Yes, we should favor composition over inheritance.

haolinw commented on src/backend/commands/vacuum.c:
```diff
@@ -2789,16 +2790,23 @@ vacuum_combine_stats(VacuumStatsContext *stats_context, CdbPgResults *cdb_pgresu
 		 * number of rel_tuples and rel_pages.
 		 */
-		pgclass_stats = (VPgClassStats *) pgresult->extras;
+		VPgClassStatsCombo *pgclass_stats_combo = (VPgClassStatsCombo *) pgresult->extras;
```
I don't think so as extras field holds multiple types of statistics structure, refer to  
```
typedef enum PGExtraType {
	PGExtraTypeNone,
	PGExtraTypeVacuumStats,		/* Stats collected for vacuum and analyze from QEs */
	PGExtraTypeTableStats		/* Table stats collected for statement from QEs */
} PGExtraType;
```

yaowangm replied:
I think we still can make some kind of type checking here such as using `union` struct. But let's keep it as is for now.

haolinw commented on src/backend/commands/vacuum.c:
```diff
@@ -2808,11 +2816,12 @@ vacuum_combine_stats(VacuumStatsContext *stats_context, CdbPgResults *cdb_pgresu
 			Assert(pgresult->extraslen == sizeof(VPgClassStats));
```
Done.

haolinw commented on src/backend/commands/vacuum.c:
```diff
@@ -2834,30 +2844,47 @@ vac_update_relstats_from_list(List *updated_stats)
+		}
+		else
+		{
+			elog(DEBUG2, "Vacuum update stats oid=%u pages=%d tuples=%f was skipped because "
```
All right. I don't feel it's serious, but maybe exposing it to log could provide some hint information.

haolinw replied:
Change back to DEBUG2 as above explanation.

haolinw commented on src/backend/commands/vacuum_ao.c:
```diff
@@ -479,14 +480,18 @@ vacuum_appendonly_indexes(Relation aoRelation, int options, Bitmapset *dead_segs
 		{
 			for (i = 0; i < nindexes; i++)
-			{
+			{	
```
Done.

haolinw commented on src/backend/commands/vacuum_ao.c:
```diff
@@ -688,7 +695,7 @@ scan_index(Relation indrel, int elevel, BufferAccessStrategy vac_strategy)
 	ivinfo.analyze_only = false;
 	ivinfo.estimated_count = true;
 	ivinfo.message_level = elevel;
```
But it is a field of IndexVacuumInfo which needs to be assigned a value anyway. And checking of this field could happen in later development (for example, upstream introduced vac_bulkdel_one_index() reporting message based on this field). Besides, `elevel` was used in later of this function, which makes me feel nothing could be removed.

soumyadeep2007 replied:
Ah good point, thanks for digging!

haolinw commented on src/test/regress/sql/uaocs_compaction/index_stats.sql:
```diff
@@ -17,3 +17,18 @@ select * from uaocs_index_stats;
 SELECT relname, reltuples FROM pg_class WHERE relname = 'uaocs_index_stats';
 SELECT relname, reltuples FROM pg_class WHERE relname = 'uaocs_index_stats_int_idx1';
+
+-- A test of index stat for access methods that rely on table tuple count (bitmap, gin)
```
Right, reworded accordingly.

haolinw commented on src/test/regress/sql/uaocs_compaction/index_stats.sql:
```diff
@@ -17,3 +17,18 @@ select * from uaocs_index_stats;
+SELECT relname, reltuples FROM pg_class WHERE relname = 'uaocs_index_stats_int_idx2';
+
+-- first vacuum collect table stat on segments
+vacuum uaocs_index_stats;
```
Done.

haolinw commented on src/test/regress/expected/uao_compaction/index_stats.out:
```diff
@@ -35,3 +35,22 @@ SELECT relname, reltuples FROM pg_class WHERE relname = 'mytab_int_idx1';
 (1 row)
 
+-- A test of index stat for access methods that rely on table tuple count (bitmap, gin)
+truncate mytab;
```
Done.

haolinw commented on src/backend/commands/vacuum_ao.c:
```diff
@@ -504,14 +509,15 @@ vacuum_appendonly_indexes(Relation aoRelation, int options, Bitmapset *dead_segs
- * since we cannot get table->reltuples in the calling context.
- * Therefore, ivinfo.num_heap_tuples is not an accurate value, so we need
- * to set estimated_count to true.
+ * We used to pass an argument rel_tuple_count with value of table->reltuples to
+ * ivinfo.num_heap_tuples, now with the new VACUUM strategy,  we cannot get
+ * exact table->reltuples in the calling context. We able to use existing
+ * value from pg_class for table. Therefore, ivinfo.num_heap_tuples is not an
+ * accurate value, so we need to set estimated_count to true.
  */
```
Done.

haolinw commented:
> (2) Is the 4th commit: `add test into btree_index.sql ` a test for the 2nd commit: `Fix incorrect index->reltuples after VACUUM `?

Yes.

haolinw commented on src/test/regress/sql/btree_index.sql:
```diff
@@ -145,3 +145,38 @@ VACUUM delete_test_table;
+-- https://github.com/postgres/postgres/commit/857f9c36cda520030381bd8c2af20adf0ce0e1d4),
+-- there was a case in QD-QEs distributed deployment that some QEs could skip full index scan and
+-- stop updating statistics, result in QD being unable to collect all QEs' stats thus overwrote
+-- a paritial accumulated value to index->reltuples. More interesting, it usually happened starting
```
The reason of why it happened on 3rd vacuum is for each vacuum, the number of left tuples will be recorded into Btree meta page, in `btm_last_cleanup_num_heap_tuples`. It is initialized as -1, the 1st vacuum will update it to 0 (without this patch), the 2nd vacuum then will update it to the current number of live tuples, next the 3rd vacuum will fail at condition checking for updating/sending statistics, see code in _bt_vacuum_needs_cleanup() below:
```
		/*
		 * If table receives enough insertions and no cleanup was performed,
		 * then index would appear have stale statistics.  If scale factor is
		 * set, we avoid that by performing cleanup if the number of inserted
		 * tuples exceeds vacuum_cleanup_index_scale_factor fraction of
		 * original tuples count.
		 */
		relopts = (StdRdOptions *) info->index->rd_options;
		cleanup_scale_factor = (relopts &&
								relopts->vacuum_cleanup_index_scale_factor >= 0)
			? relopts->vacuum_cleanup_index_scale_factor
			: vacuum_cleanup_index_scale_factor;
		prev_num_heap_tuples = metad->btm_last_cleanup_num_heap_tuples;

		if (cleanup_scale_factor <= 0 ||
			prev_num_heap_tuples <= 0 ||
			(info->num_heap_tuples - prev_num_heap_tuples) /
			prev_num_heap_tuples >= cleanup_scale_factor)
			result = true;
	}

```

haolinw replied:
It always happened.

yaowangm replied:
Pretty interesting. Make sense for me.

yaowangm reviewed (approved):
LGTM. Thanks for your work and patient explanation!

haolinw commented:
> I meant you can use VPgClassStats as a member of VPgClassStatsCombo to avoid duplication code, like:

```
typedef struct VPgClassStatsCombo
{
    VPgClassStats pgClassStats;
    int count;
}
```
I did consider implementing like this, but on side effect (or maybe not) is you need to double reference when accessing the fields in pgClassStats, such like stats->pgClassStats.rel_tuples instead of single reference and no need to bother alignment.
I would keep current pattern.

Thanks for your review @yaowangm.

Aegeaner reviewed (commented):
The happenings of the WARNING "Vacuum update stats oid=%u pages=%d tuples=%f was skipped because collected segment number %d didn't match the expected %d." seems unpredictable, we need to filter them by some regex expression in the test init file.

soumyadeep2007 commented:
For the 2nd commit: `Fix incorrect index->reltuples after VACUUM`

I strongly feel we can disable the reloption and GUC `vacuum_cleanup_index_scale_factor` from the code to get around this problem, instead of introducing net new code to **tolerate** this feature.

In fact Peter Geoghegan has done exactly that upstream. He has deleted this GUC/reloption. See: https://github.com/postgres/postgres/commit/9f3665fbfc34b963933e51778c7feaa8134ac885

Also the 2nd commit is fairly independent of the 1st commit. So here is what I suggest for this PR as a whole:

* This should only contain the 1st commit and addressed feedback for the 1st commit. Please rename the PR to relfect that.
* Open a 2nd PR that either:
    - Backport the removal of this GUC/reloption from upstream and manually verify that your repro. This will even remove the need to add a test. (**preferred**)
    - If the backport is unsuccessful in an extreme circumstance, simply disable the feature in GP. Ban the GUC/reloption, turn it off by default etc.

haolinw commented:
> For the 2nd commit: `Fix incorrect index->reltuples after VACUUM`
> 
> I strongly feel we can disable the reloption and GUC `vacuum_cleanup_index_scale_factor` from the code to get around this problem, instead of introducing net new code to **tolerate** this feature.
> 
> In fact Peter Geoghegan has done exactly that upstream. He has deleted this GUC/reloption. See: [postgres/postgres@9f3665f](https://github.com/postgres/postgres/commit/9f3665fbfc34b963933e51778c7feaa8134ac885)
> 
> Also the 2nd commit is fairly independent of the 1st commit. So here is what I suggest for this PR as a whole:
> 
> * This should only contain the 1st commit and addressed feedback for the 1st commit. Please rename the PR to reflect that.
> * Open a 2nd PR that either:
>   
>   * Backports the removal of this GUC/reloption from upstream [commit](https://github.com/postgres/postgres/commit/9f3665fbfc34b963933e51778c7feaa8134ac885) and manually verify that your repro. This will even remove the need to add a test. (**preferred**)
>   * If the backport is unsuccessful in an extreme circumstance, simply disable the feature in GP. Ban the GUC/reloption, turn it off by default etc.

The second commit doesn't address what condition should trigger btree cleanup, the real issue is, if the number of QEs returning stats is less than the number of dispatches, index->reltuples will be wrong on QD as QD blindly accumulated the received QE's stats but didn't verify the QE's number. Even if we disabled the GUC, some QE could still not return stats which is not expected. This commit force the QE returning stats always no matter whether it satisfies the condition about `vacuum_cleanup_index_scale_factor` or not.

soumyadeep2007 commented:
> The second commit doesn't address what condition should trigger btree cleanup, the real issue is, if the number of QEs returning stats is less than the number of dispatches, index->reltuples will be wrong on QD as QD blindly accumulated the received QE's stats but didn't verify the QE's number. Even if we disabled the GUC, some QE could still not return stats which is not expected. This commit force the QE returning stats always no matter whether it satisfies the condition about vacuum_cleanup_index_scale_factor or not.

Can you give an example of how without the skip logic from the GUC some QEs can omit returning the stats?

haolinw commented:
> > The second commit doesn't address what condition should trigger btree cleanup, the real issue is, if the number of QEs returning stats is less than the number of dispatches, index->reltuples will be wrong on QD as QD blindly accumulated the received QE's stats but didn't verify the QE's number. Even if we disabled the GUC, some QE could still not return stats which is not expected. This commit force the QE returning stats always no matter whether it satisfies the condition about vacuum_cleanup_index_scale_factor or not.
> 
> Can you give an example of how without the skip logic from the GUC some QEs can omit returning the stats?

Just from code logic, I think _bt_vacuum_needs_cleanup() is used to set a condition to determine whether trigger btree cleanup or not, from upstream patch you mentioned, the condition is changed but there is still a chance to returning false, right ? If returning false, current QE would skip updating and sending stats to QD, but QD does accumulation anyway, thinking if one of QE returned stats but others didn't, the final sum value would be wrong. Actually, even I force returning true in this function, there is still a chance of accumulating less QE's stats could happen. See the pipline failure issue which is happening in the pipeline that I'm debugging now, it is reporting the following message:
```
WARNING:  Vacuum update stats oid=96429 pages=2 tuples=3.000000 was skipped because collected segment number 1 didn't match the expected 3.
```

haolinw commented:
The reason why `WARNING:  Vacuum update stats oid=96429 pages=2 tuples=3.000000 was skipped because collected segment number 1 didn't match the expected 3.` was reported even forcing QE always sending stats is identified and explained at here:
```
         /*
          * We do have chance to enter this branch in the case when in compact phase.
          * For example, in compact phase, some QEs may need to drop dead segfiles,
          * while others may not. Only the QEs which dropping dead segfiles could go to
          * vacuum indexes path then update and send the statistics to QD, QD just
          * collected part of QEs' stats hence should not be as the final result to
          * overwrite QD's stats. 
          * 
          * One may think why not having the stats update only happens in the final
          * phase (POST_CLEANUP_PHASE), yes that's an alternative to get a final stats
          * exactly for QD. 
          * 
          * Given the AO/CO VACUUM is a multi-phases process which may have an interval
          * between each phase. In real circumstance, concurrent VACUUM is mostly a heavy
          * job and this interval could be significant than normal cases, hence it seems
          * better to collect and update QD's stats timely. So current strategy is, QD always 
          * collect QE's stats across phases, once we collected the expected number (means
          * same as dispatched QE number) of QE's stats, we update QD's stats subsequently,
          * instead of updating at the final phase.
          * 
          * Set the logging level as DEBUG2 as skipping sending stats here is not considered
          * a real issue.
          */
          elog(DEBUG2, "Vacuum update stats oid=%u pages=%d tuples=%f was skipped because "
	          "collected segment number %d didn't match the expected %d.", stats->relid,
	          stats->rel_pages, stats->rel_tuples, stats->count, stats_context->nsegs);
```

haolinw commented:
The pipeline failure has nothing to do with this change.

soumyadeep2007 commented:
Sorry I didn't get time to review this. Feel free to have others review it. I had one high level thought:

> Just from code logic, I think _bt_vacuum_needs_cleanup() is used to set a condition to determine whether trigger btree cleanup or not, from upstream patch you mentioned, the condition is changed but there is still a chance to returning false, right ? If returning false, current QE would skip updating and sending stats to QD, but QD does accumulation anyway, thinking if one of QE returned stats but others didn't, the final sum value would be wrong. Actually, even I force returning true in this function, there is still a chance of accumulating less QE's stats could happen. 

Yes you are right. Can we add a repro test for this without the influence of the GUC (maybe use fault injector etc)? We need to repro the problem as if the GUC isn't there in the code (if the test is dependent on the GUC, it will break once we absorb Peter's commit).

haolinw commented:
> Sorry I didn't get time to review this. Feel free to have others review it. I had one high level thought:
> 
> > Just from code logic, I think _bt_vacuum_needs_cleanup() is used to set a condition to determine whether trigger btree cleanup or not, from upstream patch you mentioned, the condition is changed but there is still a chance to returning false, right ? If returning false, current QE would skip updating and sending stats to QD, but QD does accumulation anyway, thinking if one of QE returned stats but others didn't, the final sum value would be wrong. Actually, even I force returning true in this function, there is still a chance of accumulating less QE's stats could happen.
> 
> Yes you are right. Can we add a repro test for this without the influence of the GUC (maybe use fault injector etc)? We need to repro the problem as if the GUC isn't there in the code (if the test is dependent on the GUC, it will break once we absorb Peter's commit).

Current implementation didn't consider the GUC vacuum_cleanup_index_scale_factor at all, it always does vacuum cleanup index as long as the role is QE. Even in this case, some QE still may not return stats to QD for no chance to enter index vacuum path in the temporary middle phase of vacuum, and my choice is don't treat this as a real issue, instead, update QD's stats only when received all QE's stats responses, which should be always accurate stats update to QD no matter what phase it is.

So the conclusion is:
- we allow QEs skipping sending stats to QD;
- QEs could skip sending stats to QD at needs but not depending on criteria around the GUC or later Peter's commit;
- we update QD's stats only when receiving all QEs' stats;
- we don't update QD's stats if receiving partial QEs stats responses.

I agree on backporting Peter's [commit](https://github.com/postgres/postgres/commit/9f3665fbfc34b963933e51778c7feaa8134ac885) to remove the GUC vacuum_cleanup_index_scale_factor, I think it is more appropriate to happen in another PR with related tests as well rather than this. I prefer to getting this PR to finish and then proceeding the backport in a new PR later. Thoughts ?

soumyadeep2007 commented:
> Current implementation didn't consider the GUC vacuum_cleanup_index_scale_factor at all, it always does vacuum cleanup index as long as the role is QE. Even in this case, some QE still may not return stats to QD for no chance to enter index vacuum path in the temporary middle phase of vacuum, and my choice is don't treat this as a real issue, instead, update QD's stats only when received all QE's stats responses, which should be always accurate stats update to QD no matter what phase it is.

I see, thanks for the explanation. All good then.

> So the conclusion is:
> 
> * allow QEs skipping sending stats to QD;
> * QEs could skip sending stats to QD at needs but not depending on criteria around the GUC or later Peter's commit;
> * update QD's stats only when receiving all QEs' stats;
> * don't update QD's stats if receiving partial QEs' stats.

Sanity check: We should add appropriate log messages at an appropriate log level if we are receiving partial stats and skipping etc, not sending stats from the QE etc.

> I agree on backporting Peter's [commit](https://github.com/postgres/postgres/commit/9f3665fbfc34b963933e51778c7feaa8134ac885) to remove the GUC vacuum_cleanup_index_scale_factor, I think it is more appropriate to happen in another PR with related tests as well rather than this. I prefer to getting this PR to finish and then proceeding the backport in a new PR later. Thoughts ?

Absolutely! Please take this as a next step and backport that commit after this PR, in a separate PR, to close the loop.

haolinw commented:
> Sanity check: We should add appropriate log messages at an appropriate log level if we are receiving partial stats and skipping etc, not sending stats from the QE etc.

Using LOG level to hint, which is no impact to existing tests and informative.
```
elog(LOG, "Vacuum update stats oid=%u pages=%d tuples=%f was skipped because "
         "collected segment number %d didn't match the expected %d.", stats->relid,
	 stats->rel_pages, stats->rel_tuples, stats->count, stats_context->nsegs);
```
